src/indexer/merger.rs

```diff
@@ -739,17 +797,68 @@ impl IndexMerger {
             .collect::<Vec<_>>();
 
         // We can now initialize our serializer, and push it the different values
-        let mut serialize_vals =
-            fast_field_serializer.new_u64_fast_field_with_idx(field, min_value, max_value, 1)?;
+        let stats = FastFieldStats {
+            max_value,
+            num_vals: num_vals as u64,
+            min_value,
+        };
         if let Some(doc_id_mapping) = doc_id_mapping {
-            for (doc_id, reader_with_ordinal) in doc_id_mapping {
-                let ff_reader = &fast_field_reader[reader_with_ordinal.ordinal as usize];
-                ff_reader.get_vals(*doc_id, &mut vals);
-                for &val in &vals {
-                    serialize_vals.add_val(val)?;
+            struct SortedDocidMultiValueAccessProvider<'a> {
+                doc_id_mapping: &'a Vec<(DocId, SegmentReaderWithOrdinal<'a>)>,
+                fast_field_readers: &'a Vec<MultiValuedFastFieldReader<u64>>,
+                offsets: Vec<u64>,
+            }
+            impl<'a> FastFieldDataAccess for SortedDocidMultiValueAccessProvider<'a> {
+                fn get_val(&self, pos: u64) -> u64 {
+                    // use the offsets index to find the doc_id which will contain the position.
+                    // the offsets are stricly increasing so we can do a simple search on it.
+                    let new_docid = self
+                        .offsets
+                        .iter()
+                        .position(|&offset| offset > pos)
+                        .unwrap()
```

Comment:
`expect` with some kind of message

"Pos is out of bound"... or something like that.